src/mds/CInode.cc

```diff
@@ -5176,6 +5182,111 @@ void CInode::maybe_export_pin(bool update)
   }
 }
 
+void CInode::maybe_export_ephemeral_pin(bool update)
+{
+  //If both the configs aren't set then return
+  if (!mdcache->get_export_ephemeral_distributed_config()
+      && !mdcache->get_export_ephemeral_random_config())
+    return;
+
+  double export_ephemeral_random_pin = get_export_ephemeral_random_pin(false);
+  //bool export_ephemeral_distributed_pin = get_export_ephemeral_distributed_pin();
+
+  if (update)
+    dout(10) << "I'm here for the update" << dendl;
+
+  //Check if it's already ephemerally pinned
+  if ((is_export_ephemeral_random_pinned || is_export_ephemeral_distributed_pinned) && !update)
+      return;
+
+  if (mdcache->get_export_ephemeral_random_config()) {
+    if ((update || export_ephemeral_random_pin >=
+	  ceph::util::generate_random_number(0.0, 1.0))
+	&& is_export_ephemeral_distributed_pinned == false) {
+
+      dout(10) << "I'm here under ephemeral random because is_export_ephemeral_distributed is" << is_export_ephemeral_distributed_pinned << dendl;
+
+      is_export_ephemeral_random_migrating = true;
+
+      bool queue = false;
+      for (auto p = dirfrags.begin(); p != dirfrags.end(); p++) {
+        CDir *dir = p->second;
+        if (!dir->is_auth())
+          continue;
+        if (dir->is_subtree_root()) {
+	  // set auxsubtree bit or export it
+          if (!dir->state_test(CDir::STATE_AUXSUBTREE) ||
+            mdcache->hash_into_rank_bucket(ino(), mdcache->mds->mdsmap->get_max_mds()) != dir->get_dir_auth().first)
+            queue = true;
+        } else {
+        // create aux subtree or export it
+        queue = true;
+        }
+        if (queue) {
+	  if (mdcache->hash_into_rank_bucket(ino(), mdcache->mds->mdsmap->get_max_mds()) == mdcache->mds->get_nodeid()) {
```

Comment:
what does this mean?

Reply:
Each rank has its own data structure (consistent_hash_inodes) to keep a track of the inodes that it handles as part of the consistent hash ring. If the inode no hashes to the same rank as the current mds (i.e no migration is required), then insert it into the data structure. The data structure is else populated during migration on the importer's side during import.

Reply:
what does 'mdcache->hash_into_rank_bucket(ino(), ....' means

Reply:
This method hashes the inode no into a rank.. almost according to a consistent hashing strategy specified in this paper: https://arxiv.org/pdf/1406.2294.pdf (Jump Consistent Hash)